torch/csrc/jit/passes/common_subexpression_elimination.cpp

```diff
@@ -0,0 +1,145 @@
+#include "torch/csrc/jit/ir.h"
+
+#include <algorithm>
+#include <unordered_map>
+
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/jit/passes/common_subexpression_elimination.h"
+
+namespace torch { namespace jit {
+
+
+
+// Check whether two nodes have the same attributes in CSE.
+// This function may be too conservative for general use.
+// Do NOT support t/ts/g/gs attributes.
+// If t/ts are supported, CONSTANT node comparison may need to consider device.
+bool attributesEqualCSE(const Node* lhs, const Node* rhs) {
+  JIT_ASSERT(lhs != nullptr);
+  JIT_ASSERT(rhs != nullptr);
+  // One has attributes, the other does not.
+  if (lhs->hasAttributes() != rhs->hasAttributes()) return false;
+  // Neither has attributes.
+  if (!lhs->hasAttributes() && !rhs->hasAttributes()) return true;
+
+  auto lnames = lhs->attributeNames();
+  auto rnames = rhs->attributeNames();
+  if (lnames != rnames) return false;
+
+  Node* l = const_cast<Node*>(lhs);
```

Comment:
Blah, which operations needed to be const but weren't? Just curious.

Reply:
Looks like these casts are not necessary now. Remove them :-)